Inspector : Add support for disabling edits

murraystevenson opened the pull request:
This moves the potentially Inspector-specific "Disable Edit" functionality from InspectorColumn to the Inspectors themselves via the new `Inspector::Result::disableEdit()`. Inspectors with custom disabling logic, such as the SetMembershipInspector, can provide their own by implementing `disableEditFunction()`.

On the user-facing side, this PR allows toggling of edits disabled via the "Disable Edit" actions in the Light Editor and Render Pass Editor. The approach taken is to only allow edits disabled in the current session to be re-enabled to allow for investigative "how would my scene be affected by disabling this edit?" toggling, while attempting to avoid enabling edits the user may not expect to exist, such as previously unmodified Spreadsheet cells in EditScope processors.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -176,43 +176,18 @@ def __disablableInspectionTweaks( pathListing ) :
 
 			with inspectionContext :
 				inspection = column.inspector().inspect()
-				if inspection is not None and inspection.editable() :
```
I'm not sure how this is happening, but the right-click menu for a read-only edit is emitting this error in the shell and showing an incomplete menu rather than showing a greyed-out "Disabled Edit" item :

```ERROR : Emitting signal : Not editable : RectLight.visualiserAttributes.scale is locked.```

That's editing the scale on the light below :

<details>

```
import Gaffer
import GafferScene
import GafferUSD
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 5, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 0, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["RectLight"] = GafferUSD.USDLight( "RectLight" )
parent.addChild( __children["RectLight"] )
__children["RectLight"].loadShader( "RectLight" )
__children["RectLight"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["RectLight"]["name"].setValue( 'rectLight' )
__children["RectLight"]["parameters"]["arnold:aov"]["enabled"].setValue( True )
__children["RectLight"]["parameters"]["shaping:cone:angle"]["value"].setValue( 25.0 )
__children["RectLight"]["visualiserAttributes"]["scale"]["enabled"].setValue( True )
Gaffer.Metadata.registerValue( __children["RectLight"]["visualiserAttributes"]["scale"], 'readOnly', True )
__children["RectLight"]["__uiPosition"].setValue( imath.V2f( -5.75, 9.85000038 ) )


del __children

```

</details>

johnhaddon replied:
Likewise, the `D` shortcut is emitting this message to the shell rather than showing a pop-up message :

```ERROR : EventSignalCombiner : Not editable : RectLight.visualiserAttributes.scale is locked.```

murraystevenson replied:
Thanks for finding that one, I'm now properly elevating that `RectLight.visualiserAttributes.scale is locked.` message to the user in the form of a pop-up as part of the overall messaging improvements in 1968cbf.

johnhaddon commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -296,6 +297,8 @@ void Inspector::inspectHistoryWalk( const GafferScene::SceneAlgo::History *histo
 							result->m_editFunction = nonEditableReason;
 						}
+
+						result->m_disableEditFunction = disableEditFunction( result->m_source.get(), history );
```
I think we're making the right decision in not allowing edits to be disabled unless they are in the current EditScope. The guarantee that the UI won't touch anything outside of the scope is a fundamental one.

But at present it's not obvious why something can't be disabled when it's outside the scope - it took me a little while to figure out in my first interactive testing for this PR, even though I was familiar with the rules from before. At the moment, `m_disableEditFunction` remains uninitialised in this case, but perhaps we can initialise it with a message describing the reason it can't be disabled? I'm pretty sure we guarantee that `m_editFunction` is always initialised, and it seems like the same might be good for `m_disableEditFunction`.

murraystevenson replied:
I've taken a stab at describing as many of the non-disablable reasons as part of the overall messaging improvements in 1968cbf. Some of the wording could potentially be improved, suggestions very welcome there, I've tried to keep things as terse as possible given the ephemeral nature of the popups...

johnhaddon commented on include/GafferSceneUI/Private/Inspector.h:
```diff
@@ -343,6 +351,13 @@ class GAFFERSCENEUI_API Inspector::Result : public IECore::RefCounted
 		/// by `acquireEdit()`. This should be displayed to the user.
 		std::string editWarning() const;
 
+		/// Returns `true` if `disableEdit()` will disable the edit
+		/// at `source()`, and `false` otherwise.
+		bool canDisableEdit() const;
+		/// Disables the edit at `source()`. Throws if
+		/// `!canDisableEdit()`
+		void disableEdit() const;
```
We seem to be lacking the equivalent of `nonEditableReason()` here - the only way to find out why you can't disable an edit is to call `disableEdit()` and catch the exception.

murraystevenson replied:
Ah yes, that would be useful and is necessary for the overall messaging improvements. Added in 6e14602.

After writing it, `nonDisablableReason()` still feels a bit clumsy to me, but fits with `nonEditableReason()`...

johnhaddon replied:
Since it's an imaginary word, I suppose we can spell it how we want, but should it be `nonDisableableReason()` (extra e)? No matter how we spell it, I suspect I will always read it as "nonBlaBlaReason" :(

My only other thought was to pass a `string *reasonNot` to `canDisableEdit()`, but that gets awkward with the Python bindings. I think we can live with the awkward name...

murraystevenson replied:
Changed to `nonDisableableReason()` and squashed.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -176,18 +177,57 @@ def __disablableInspectionTweaks( pathListing ) :
+					if isinstance( edit, ( Gaffer.NameValuePlug, Gaffer.OptionalValuePlug, Gaffer.TweakPlug ) ) and Gaffer.Metadata.value( edit, "inspector:disabledEdit" ) :
+						disabledTweaks.append( inspection )
+					else :
+						# The selection includes a cell that is not viable candidate for enabling or disabling,
+						# we don't allow toggling of any of the selected cells in this case.
+						return [], True
```
Oh, looks like you've changed the "silently ignore the others" behaviour I was questioning earlier. That's good I think. But it seems we've thrown away information about why something couldn't be disabled by doing that on this branch. Ideally we'd surface as much of that information to the user as possible when they hit `D` and nothing happens. And perhaps we could also include it in a tooltip for a greyed-out "Disable Edit" menu item?

murraystevenson replied:
Now surfacing as much info as I have to hand as part of the overall messaging improvements in 1968cbf.

Including a tooltip on a menu item via `"description"` exposed an issue where we weren't hiding the tooltip when the cursor moved to a menu item without one, so I've addressed that separately in d09d2a6.

johnhaddon commented on python/GafferSceneUI/_InspectorColumn.py:
```diff
@@ -360,11 +400,12 @@ def __contextMenu( column, pathListing, menuDefinition ) :
+	tweaks, tweaksEnabled = __toggleableInspectionTweaks( pathListing )
 	menuDefinition.append(
-		"Disable Edit",
+		"{} Edit{}".format( "Disable" if tweaksEnabled else "Enable", "s" if len( tweaks ) > 1 else "" ),
```
Perhaps "Reenable" rather than "Enable"? Maybe that suggests something that was recently disabled a little better?

murraystevenson replied:
Switched to "Reenable" in 8368682.

johnhaddon reviewed (commented):
Thanks Murray!

API-wise this is feeling pretty good to me, but I've noted a couple of issues in the way we're surfacing it to the user. One I think is just a bug that I haven't been able to locate in the code, and the other is just a request for more pop-ups explaining the reasons behind `D` not working. Hopefully those make sense...

Cheers...
John

murraystevenson commented on src/GafferSceneUIModule/InspectorBinding.cpp:
```diff
@@ -72,10 +72,10 @@ IECore::ObjectPtr valueWrapper( const GafferSceneUI::Private::Inspector::Result
 }
 
-Gaffer::ValuePlugPtr acquireEditWrapper( GafferSceneUI::Private::Inspector::Result &result )
+Gaffer::ValuePlugPtr acquireEditWrapper( GafferSceneUI::Private::Inspector::Result &result, bool createIfNecessary = true )
```
Makes sense! I've tidied that up in d017641.

murraystevenson commented on src/GafferSceneUI/SetMembershipInspector.cpp:
```diff
@@ -194,6 +173,30 @@ bool SetMembershipInspector::editSetMembership( const Result *inspection, const
+	Metadata::nodeValueChangedSignal().connect( boost::bind( &SetMembershipInspector::nodeMetadataChanged, this, ::_2, ::_3 ) );
+}
+
+bool SetMembershipInspector::editSetMembership( const Result *inspection, const ScenePlug::ScenePath &path, EditScopeAlgo::SetMembership setMembership ) const
```
Correct, we should be free to remove `editSetMembership()` from the public API once we have the direct editing Inspector API in place...

murraystevenson commented on src/GafferSceneUI/SetMembershipInspector.cpp:
```diff
@@ -319,6 +322,24 @@ Inspector::EditFunctionOrFailure SetMembershipInspector::editFunction( Gaffer::E
+	else
+	{
+		return [
+			plug,
```
Thanks for pointing that out, now capturing a `PlugPtr` in 0de3c7b.

murraystevenson commented:
Thanks for the input! I've addressed the comments inline and taken an overall run at improving the user messaging when disabling isn't possible. Might now be erring on the side of too much info, but we should now be covering the common cases...

johnhaddon commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -291,14 +296,25 @@ void Inspector::inspectHistoryWalk( const GafferScene::SceneAlgo::History *histo
+					else if( node->ancestor<EditScope>() && node->ancestor<EditScope>() != result->m_editScope )
+					{
+						result->m_disableEditFunction = fmt::format(
+							"Edit authored in EditScope ({}), set it as the target EditScope to disable.",
```
This isn't unique to this PR, but I find messages of this form are often hard to read because my EditScope is often called EditScope, leading to a sentence full of EditScope. Maybe that's just because I'm throwing scenes together for testing, and everyone else has better names? But I wonder if we should refer to "the scope in which you have chosen to edit" as the "edit scope" rather than the "EditScope", to help with this?

I also wonder if there could be clarity to be gained by relegating the changeable details to a second sentence, so the first sentence is always the same, and always very blunt.

> Edit is not in the current edit scope. Change scope to \<NameOfEditScope\> to disable.

murraystevenson replied:
Switched over to the suggested `Edit is not in the current edit scope. Change scope to <NameOfEditScope> to disable.`

Also took the opportunity to conform the Inspector's existing "EditScope (EditScope)" messages to the same style in c4ddc45.

johnhaddon commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -291,14 +296,25 @@ void Inspector::inspectHistoryWalk( const GafferScene::SceneAlgo::History *histo
+					}
+					else if( !node->ancestor<EditScope>() && result->m_editScope )
+					{
+						result->m_disableEditFunction = "Edit was not authored in any EditScope, set target EditScope to (None) to disable.";
```
If we went with my suggestion above, I think this could follow the same form :

> Edit is not in the current edit scope. Change scope to None to disable.

murraystevenson replied:
Switched over to the suggested `Edit is not in the current edit scope. Change scope to None to disable.`

johnhaddon reviewed (commented):
Thanks Murray, LGTM! I've made a couple of suggestions for alternative wording inline - feel free to take or leave them. Either way, let's get this squashed and merged...

murraystevenson commented on src/GafferSceneUI/Inspector.cpp:
```diff
@@ -212,18 +212,23 @@ Inspector::ResultPtr Inspector::inspect() const
+			// the property.
+			result->m_editFunction = "No editable source found in history.";
+		}
+		result->m_disableEditFunction = "No disablable source found in history.";
```
Switched to "No editable source found in history" and squashed it in.
